cumulusci/tasks/salesforce/users/permsets.py

```
@@ -29,65 +31,98 @@ def _init_options(self, kwargs):
+
+        missing_perms = [
+            api_name
+            for api_name in self.options["api_names"]
+            if api_name not in perms_by_ids.values()
+        ]
+        if missing_perms:
+            raise CumulusCIException(
+                f"The following {self.permission_label}s were not found: {', '.join(missing_perms)}."
+            )
+        return perms_by_ids
+
+    def _get_assignments(self, user_id, assigned_perms):
+        assignments = []
+        for perm, perm_name in self.perms_by_id.items():
+            if perm not in assigned_perms:
+                self.logger.info(
+                    f'Assigning {self.permission_label} "{perm_name}" to {user_id}.'
                 )
-
-        # Assign all not already assigned
-        for api_name in self.options["api_names"]:
-            if perms[api_name] not in assigned_perms:
-                self.logger.info(f'Assigning {self.permission_label} "{api_name}".')
                 assignment = {
-                    "AssigneeId": user["Id"],
+                    "attributes": {"type": self.assignment_name},
+                    "AssigneeId": user_id,
+                    self.assignment_lookup: perm,
                 }
-                assignment[self.assignment_lookup] = perms[api_name]
-                # Create the new assignment.
-                getattr(self.sf, self.assignment_name).create(assignment)
+                assignments.append(assignment)
             else:
                 self.logger.warning(
-                    f'{self.permission_label} "{api_name}" is already assigned.'
+                    f'{self.permission_label} "{perm_name}" is already assigned to {user_id}.'
                 )
+        return assignments
+
+    def _insert_assignments(self, records_to_insert):
+        request_body = json.dumps({"allOrNone": False, "records": records_to_insert})
+        result = self.sf.restful("composite/sobjects", method="POST", data=request_body)
+        self.logger.debug(result)
```

This output seems a little user-unfriendly. Could we just throw an exception with a specific message if any subrequest fails?